### PR TITLE
Deprecate Enum.uniq/2 in favor of Enum.uniq_by/2

### PR DESCRIPTION
**Pull request by shankardevy**
Fixes https://github.com/elixir-lang/elixir/issues/3142

### CODE REVIEW

**Comment by josevalim** [comment]
Great! Can you please port the tests that use Enum.uniq with two arguments to Enum.uniq_by? Thank you!

**Comment by michalmuskala** [comment]
Shouldn't uniq_by not use default fun?

**Comment by josevalim** [comment]
Good catch @michalmuskala. uniq_by should not have a default function.

**Comment by shankardevy** [comment]
on another note, can someone shed light on why `uniq/2` needs to be renamed to `uniq_by/2`? There is no related discussion on irc or on google groups that I could find.

http://irc.programmingelixir.com/search/keyword?q=Enum.uniq (IRC history search of elixir channel since 2013)
https://groups.google.com/forum/#!searchin/elixir-lang-talk/Enum.uniq

**Comment by josevalim** [comment]
It is being renamed so it becomes consistent with max_by, sort_by and friends.

File: lib/elixir/lib/enum.ex [hunk 1]
```diff
@@ -2046,6 +2048,27 @@ defmodule Enum do
...
+
+      iex> Enum.uniq_by([{1, :x}, {2, :y}, {1, :z}], fn {x, _} -> x end)
+      [{1,:x}, {2,:y}]
+
+  """
+  @spec uniq_by(t, (element -> term)) :: list
+
+  # TODO: Use a HashSet
+  def uniq_by(collection, fun) when is_list(collection) do
+    do_uniq(collection, [], fun)
```
Comment (josevalim) [inline lib/elixir/lib/enum.ex]: This should use HashSet.new now (according to master). We can also remove the TODO.

**Comment by josevalim** [comment]
@shankardevy can you please rebase, squash the commits and change the tests to use Enum.uniq_by/2? Thank you!

**Comment by shankardevy** [comment]
@josevalim I have rebased, squashed commits and added tests. Thanks

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by shankardevy** [comment]
Wow! my first commit to elxir. Feeling excited. Thanks @josevalim and the community for all the mentoring all the way.

**Pull request merged**